platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,343 @@
+Today best practice is to not use labels with a `null` value.
+
+The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:
+
+```java
+interface Metric {
+    
+    //...
+    
+    @NonNull
+    Set<Label> getLabels();
+}
+```
+
+Like the name or category, the labels are defined when a metric is created and cannot be changed later.
+Based on that the MetricConfig class will be extended by a new method called `withLabels`:
+
+```java
+abstract class MetricConfig {
+    
+    //...
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Label... labels) {...}
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Set<Label> labels) {...}
+}
+```
+
+By adding labels the unique identifier of a metric is the name and the labels.
+Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:
+
+```java
+
+interface Metrics {
+    
+    // old methods:
+    // Object getValue(@NonNull String category, @NonNull String name);
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+    
+    void remove(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+}
+```
+
+Since the old methods are only used in tests or demos today we can do that change easily.
+
+> [!NOTE]  
+Since most of the filter methods that the `Metrics` interface defines today are not used we will not introduce any additional filter methods in this proposal.
+If we see that additional methods are needed we can add them later.
+
+We will have labels that are global for a `Metrics` instance.
+The best example is the `nodeId` label.
+We do not want to add the `nodeId` label to the creation of each metric.
+Instead of that we want to add the `nodeId` label to the `Metrics` instance.
+That should happen when a `Metrics` instance is created.
+The creation of a `Metrics` instance is not public api today.
+With the implementation of this proposal a `MetricsFactory` interface should be introduced that allows to create a `Metrics` instance.
+The `MetricsFactory` interface will be defined like this:
+
+```java
+interface MetricsFactory {
+    
+    @NonNull
+    default Metrics create() {
+        return create(Set.of());
+    }
+    
+    @NonNull
+    Metrics create(@NonNull Set<Label> defaultLabels);
+}
+```
+
+The definition of the api should follow our rules regarding services as defined at https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/base/service-architecture/service-architecture.md
+
+##### Examples
+
+Here are some examples how to define a metric with labels:
+
+```java
+final Label transactionTypeLabel = new Label("transactionType", "fileUpload");
+final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels(transactionTypeLabel);
```

Comment:
I would prefer a solution that does not require me to create `Label` instances, e.g.:
```
final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount")
    .addLabel("transactionType", "fileUpload");
```

Reply:
Again both might be usefull. Will add that